include/stdexec/execution.hpp

```diff
@@ -3103,76 +3096,123 @@ namespace stdexec {
+          //   1. It was already removed by another thread, or
+          //   2. It hasn't been added yet (see `start` below), or
+          //   3. The underlying operation has already completed.
+          //
+          // In each case, the right thing to do is nothing. If (1) then we raced with another
+          // thread and lost. In that case, the other thread will take care of it. If (2) then
+          // `start` will take care of it. If (3) then this stop request is safe to ignore.
+          if (!__sh_state_->__waiters_.remove(this))
+            return;
+        }
+
+        // The following code and the __notify function cannot both execute. This is because the
+        // __notify function is called from the shared state's __notify_waiters function, which
+        // first sets __waiters_ to the completed state. As a result, the attempt to remove `this`
+        // from the waiters list above will fail and this stop request is ignored.
+        __release_shared_state();
+        stdexec::set_stopped(static_cast<_Receiver&&>(this->__receiver()));
+      }
+
+      // This is called from __shared_state::__notify_waiters when the input async operation
+      // completes; or, if it has already completed when start is called, it is called from start:
+      // __notify cannot race with __on_stop_request. See comment in __on_stop_request.
       template <class _Tag>
-      static void __action(__local_state_base* __self, __action_kind __kind) noexcept {
-        auto* const __op = static_cast<__local_state*>(__self);
-        if (__kind == __action_kind::__notify) {
-          __op->__on_stop_.reset();
-
-          // The split algorithm sends by T const&. ensure_started sends by T&&.
-          if constexpr (same_as<__split::__split_t, _Tag>) {
-            std::visit(
-              __notify_visitor(__op->__receiver()), std::as_const(__op->__shared_state_->__data_));
-          } else {
-            std::visit(
-              __notify_visitor(__op->__receiver()), std::move(__op->__shared_state_->__data_));
-          }
-        } else {
-          // This is a detach operation
-          if constexpr (same_as<__split::__split_t, _Tag>) {
-            // no-op
-          } else {
-            __op->__shared_state_->__detach();
-          }
+      static void __notify(__local_state_base* __base) noexcept {
+        auto* const __self = static_cast<__local_state*>(__base);
+
+        // The split algorithm sends by T const&. ensure_started sends by T&&.
+        constexpr bool __is_split = same_as<__split::__split_t, _Tag>;
+        using __variant_t = decltype(__self->__sh_state_->__results_);
+        using __cv_variant_t = __if_c<__is_split, const __variant_t&, __variant_t>;
+
+        __self->__on_stop_.reset();
+
+        // Copy this into a local because the call to std::visit can cause __self to be deleted.
+        auto* __sh_state = __self->__sh_state_;
+        bool const __holds_ref_ = __self->__holds_ref_.exchange(false);
```

Comment:
Naming: local vars shouldn't have trailing `_` - shadows member.

Reply:
this was the accidental result of a bad search and replace